Added Cami Opal - A tool for evaluating taxonomic metagenome profilers

Alby-Git opened the pull request:
### Tool details for Cami Opal
- **Name:** Cami Opal
- **Description:** A tool to assess taxonomic metagenome profilers, predicting and evaluating the presence and relative abundance of microorganisms in microbial community DNA samples using metrics like Unifrac error and L1-norm error, and visual representations for these data.
- **Homepage:** [Cami Opal](https://github.com/CAMI-challenge/OPAL)
- **License:** [License](https://github.com/CAMI-challenge/OPAL/blob/master/LICENSE)

### Checklist for submission
**FOR CONTRIBUTOR:**
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [x] - This PR adds a new tool or tool collection
* [ ] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

I look forward to any feedback and thank you for your consideration.

bernt-matthias commented on tools/cami-opal/cami_opal.xml:
```diff
@@ -0,0 +1,272 @@
+            #if $html_output == 'true'
+            ## Copy the results to the specified output folder and suppress error outputs
+                && cp output/results.html '${html}' 2> /dev/null 
+                && cp -r output/* '$htmloutputfolder'
```
What is in this folder? If its images, etc that is used in the html file it should be copied to `$html.extra_files_path` (check for instance the dexseq tool in this repo).

bernt-matthias commented on tools/cami-opal/cami_opal.xml:
```diff
@@ -0,0 +1,272 @@
+            help="Input the gold standard file here. Recommended format: CAMI Profiling Bioboxes or BIOM format." />
+        <repeat name="input_files" title="Input files"
+            help="Enter multiple input files. Recommended format: CAMI Profiling Bioboxes or BIOM format.">
+            <param name="file" format="tabular" type="data" label="Input file" />
```
you can allow `biom1` or `biom2` in the format?

paulzierep replied:
FROM: https://github.com/CAMI-challenge/OPAL: `Note: Support for the BIOM format has been dropped (temporarily) in OPAL 1.0.4 due to incompatibility with Python 3.7.*. ` 
So I think for this version not. But you could adapt the help.

bernt-matthias commented on tools/cami-opal/cami_opal.xml:
```diff
@@ -0,0 +1,272 @@
+    <inputs>
+        <param name="gold_standard_file" format="tabular" type="data" label="Gold standard file" 
+            help="Input the gold standard file here. Recommended format: CAMI Profiling Bioboxes or BIOM format." />
+        <repeat name="input_files" title="Input files"
```
Instead of a repeat I would prefer to use `multiple="true"` for the data param.

bernt-matthias commented on tools/taxonkit/profile2cami/profile2cami.xml:
```diff
@@ -0,0 +1,111 @@
+    <tests>
+        <!-- Test 1: Basic functionality with default parameters -->
+        <test expect_num_outputs="1">
+            <param name="input_file" value="abundance.tsv" ftype="tsv" />
```
Is the cami format well defined? Maybe we can add a special Galaxy datatype for it?

bernt-matthias replied:
And I guess the taxonkit stuff should be only in https://github.com/galaxyproject/tools-iuc/pull/6085

Alby-Git replied:
Thank you @bernt-matthias and @paulzierep for pointing out the inclusion of the `taxonkit` changes in this PR. I inadvertently included commits related to `taxonkit` here which should have only been part of #6085.

I'll submit a new PR with the corrected changes shortly.

Alby-Git replied:
> Is the cami format well defined? Maybe we can add a special Galaxy datatype for it?

How would we implement a special Galaxy datatype for the CAMI  format?

bernt-matthias replied:
This would need to go to the Galaxy core. But the main question is if its well defined. And if we get much from this.

bernt-matthias commented on tools/taxonkit/profile2cami/.shed.yml:
```diff
@@ -0,0 +1,10 @@
+name: profile2cami
```
I would suggest to put the tool directly in the taxonkit folder (i.e. no sub folder) and make it a suite.

Alby-Git replied:
I can follow your advice to move the tool there, but could you please clarify what you mean by "make it a suite"?

bernt-matthias replied:
A suite is a collection of tools that belong together, e.g. the dada2 tools: https://github.com/galaxyproject/tools-iuc/blob/main/tools/dada2/.shed.yml 

The shed yaml automatically defines an entry for the toolshed for each tool and also one suite which contains all tools.

bernt-matthias reviewed (commented):
Some comments inline. 

Out of interest: What is the difference or connection to https://github.com/galaxyproject/tools-iuc/tree/main/tools/cami_amber

paulzierep commented:
> Some comments inline.
> 
> Out of interest: What is the difference or connection to https://github.com/galaxyproject/tools-iuc/tree/main/tools/cami_amber

amber is the cami binning evaluation tool (you need to provide a gold standard bin assignment / and optionally taxonomy for each bin) - so for DAS TOOL or other binners, opal is the cami evaluation tool for taxonomic composition - so Kraken2 / MetaPhlan ... in theory OPAL should also work for amplicon data (in terms of taxonomic assignment benchmarking) ... another thing to look into ....

bernt-matthias commented on tools/cami-opal/cami_opal.xml:
```diff
@@ -0,0 +1,272 @@
+        <param name="labels" type="text" value="" optional="true"
+            label="Comma-separated profiles names"
+            help="Comma-separated profiles names" />
+        <param argument="--time" type="text" value="" optional="true"
+            label="Comma-separated runtimes in hours"
+            help="Comma-separated runtimes in hours" />
+        <param argument="--memory" type="text" value="" optional="true"
+            label="Comma-separated memory usages in gigabytes"
+            help="Comma-separated memory usages in gigabytes" />
```
I see. So the two parameters are supposed to list the runtimes and used memory that have been used to compute the profiles? 

My first question is if this will be used? Technically you can find this out in Galaxy (manually), but still the question is if this is useful information.

If so, it might be an option to keep the repeat and add a runtime and memory parameter in the repeat. Then the assignment between profile, memory, and time is clearer.

Alby-Git replied:
The `-t/--time` and `-m/--memory` parameters in `opal.py` require manual entry of runtimes and memory usage, which Galaxy already tracks automatically - am I right? I find these manual entries thereby a bit redundant. Do you think we still need to implement this function?

paulzierep replied:
I would allow adding it as optional, one could get this info from galaxy and add manually

Alby-Git commented:
I apologize for the confusion caused by the multiple commits in this pull request. I have addressed the issues and made the necessary adjustments to the code. Please take another look and provide any further feedback.

Thank you for your assistance!

Best regards

paulzierep reviewed (commented):
Can u also fix @bernt-matthias comments ?

paulzierep commented on tools/cami-opal/cami_opal.xml:
```diff
@@ -0,0 +1,266 @@
+            -o output
+    
+            #if $html_output
+                ## Copy the results to the specified output folder and suppress error outputs
```
```suggestion
                ## Copy the results to the specified output folder
```

paulzierep commented on tools/cami-opal/cami_opal.xml:
```diff
@@ -0,0 +1,269 @@
+            truevalue="--normalized_unifrac" falsevalue="" />
+    </inputs>
+    <outputs>
+        <data name="archive_output" format="tar" from_work_dir="output.tar.gz" label="${tool.name} on ${on_string}: tar.gz of the complete output">
```
can we also get the tool and rank specific output csvs as collection each

paulzierep commented on tools/cami-opal/cami_opal.xml:
```diff
@@ -0,0 +1,269 @@
+        <param name="gold_standard_file" type="data" format="data" label="Gold standard file"
+            help="Input the gold standard file here. Format: CAMI Profiling Bioboxes. Note: BIOM format support is temporarily dropped in OPAL 1.0.4." />
+        <param name="input_files" type="data" format="data" multiple="true" label="Input files"
+            help="Enter multiple input files. Format: CAMI Profiling Bioboxes. Note: BIOM format support is temporarily dropped in OPAL 1.0.4." />
```
and so far still is :)

paulzierep commented on tools/cami-opal/cami_opal.xml:
```diff
@@ -0,0 +1,269 @@
+            help="Normalize the samples to compare them on the same scale."
+            truevalue="-n" falsevalue="" />
+        <param argument="--filter" type="integer" value="" optional="true"
+            label="Filter out the smallest relative abundances summing up to FILTER% within a rank"
```
can you explain better, what is the meaning of FILTER%

paulzierep replied:
could you add min / max values as well (should it be 0-1 or 0-100 ??)

paulzierep commented on tools/cami-opal/cami_opal.xml:
```diff
@@ -0,0 +1,269 @@
+            label="Plot abundances in the gold standard"
+            help="Plot abundances in the gold standard (can take some minutes)"
+            truevalue="-p" falsevalue="" />
+        <param name="labels" type="text" value="" optional="true"
```
maybe rather use a repeat, and explain, that we need as many inputs as we gt input files, I think that could even be checked with a validator

paulzierep replied:
one repeat for each file + name would be ideal

bernt-matthias replied:
I do not agree 100%. Might be nice for using the tool interactively. But it does not work in workflows, since then the number of repeat elements is fixed in the workflow.

bernt-matthias replied:
Use collection input and the collection element identifiers as labels?

paulzierep replied:
agreed !

paulzierep reviewed (commented):
minor changes please

paulzierep commented on tools/cami_opal/cami_opal.xml:
```diff
@@ -0,0 +1,290 @@
+            <filter>html_output</filter>
+        </data>
+        <data name="result" format="tabular" from_work_dir="output/results.tsv" label="${tool.name} on ${on_string}: Results" />
+        <collection name="rank_output" type="list" label="${tool.name}: Rank tables" format="tabular" >
```
The output for both collections should be tabular

paulzierep commented on tools/cami_opal/cami_opal.xml:
```diff
@@ -0,0 +1,290 @@
+            help="Normalize the samples to compare them on the same scale."
+            truevalue="-n" falsevalue="" />
+        <param argument="--filter" type="integer" value="0" optional="true"
+            label="Filter out the smallest relative abundances summing up to FILTER% within a rank"
```
Can you clarify this label ?

paulzierep commented on tools/cami_opal/cami_opal.xml:
```diff
@@ -0,0 +1,290 @@
+            truevalue="true" falsevalue="false" checked="true" />
+        <param argument="-n" name="normalize" type="boolean" optional="true"
+            label="Normalize samples" 
+            help="Normalize the samples to compare them on the same scale."
```
Can you add an test case where normalization is shown, the 3 examples are all normalized already afaik.

bgruening replied:
is this comment resolved?

paulzierep commented on tools/cami_opal/.shed.yml:
```diff
@@ -0,0 +1,12 @@
+name: cami_opal
+owner: iuc
+description: Evaluation package for metagenome binning and taxonomic assignments
```
I don't think it is used for binning

paulzierep commented on tools/cami_opal/cami_opal.xml:
```diff
@@ -0,0 +1,290 @@
+            - This file is essential for the evaluation and should be in BIOM format.
+
+        2. **Profiles Files**
+            - Multiple profiles files are required for evaluation. These files should be in CAMI Profiling Bioboxes format or BIOM format.
```
Biom does not work as we know :) can you also mention profile2cami to make it easier for the user

paulzierep reviewed (commented):
Minor changes please

paulzierep commented on tools/cami_opal/cami_opal.xml:
```diff
@@ -0,0 +1,290 @@
+            <param name="output_collections" value="false"/>
+            <param name="normalize" value="false"/>
+            <output name="result" ftype="tabular">
+                <assert_contents>
```
here you could also compare the output with the expected output instead of has text

paulzierep commented on tools/cami_opal/macros.xml:
```diff
@@ -5,7 +5,7 @@
             <yield/>
         </requirements>
     </xml>
-    <token name="@TOOL_VERSION@">1.0.12</token>
```
There seems to be an issue with the updated version

paulzierep replied:
I don't think you can specify the build version, it will take the newest build.

paulzierep commented on tools/cami_opal/macros.xml:
```diff
@@ -0,0 +1,17 @@
+<macros>
```
can you add bio.tools

paulzierep commented on tools/cami_opal/cami_opal.xml:
```diff
@@ -0,0 +1,304 @@
+        1. **HTML Report**: An HTML file containing visualizations and summary of the evaluation.
+        2. **Results File**: A TSV file with detailed evaluation metrics for each profile.
+
+        **Additional Information**
```
This is not needed, in the wrapper it will run :)

Alby-Git replied:
Right :)

paulzierep reviewed (commented):
Minor changes please, then good from my side

paulzierep commented on tools/cami_opal/cami_opal.xml:
```diff
@@ -0,0 +1,301 @@
+<tool id="cami_opal" name="CAMI OPAL" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="@PROFILE@">
+    <description>Evaluation tool for metagenome benchmark datasets,
```
Evaluation tool for multiple read-based metagenomic taxonomic profilers.

paulzierep reviewed (commented):
maybe a catch description, then good to go

bgruening commented on tools/cami_opal/cami_opal.xml:
```diff
@@ -0,0 +1,300 @@
+<tool id="cami_opal" name="CAMI OPAL" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="@PROFILE@">
+    <description>Evaluation tool for multiple read-based metagenomic taxonomic profilers.</description>
```
```suggestion
    <description>Evaluation tool for multiple read-based metagenomic taxonomic profilers</description>
```

bgruening commented on tools/cami_opal/cami_opal.xml:
```diff
@@ -0,0 +1,300 @@
+                'inputs/${safe_identifier}'
+            #end for
+
+            #set $sep = ''
```
Can you use something like `','.join($labels)` here?

bgruening commented on tools/cami_opal/cami_opal.xml:
```diff
@@ -0,0 +1,300 @@
+                ## Copy the results to the specified output folder
+                && mkdir '$htmlreport.extra_files_path'
+                && cp output/results.html $htmlreport
+                && cp -r output/* '$htmlreport.extra_files_path'
```
Do all output files needs to be in this folder or only a handful that are part of the HTML?

bgruening replied:
How large is such a folder?

Albert-Ber replied:
For my benchmarks, the folder was approximately 32.8 kB.

Albert-Ber replied:
I checked the detailed examples form: [Here](https://github.com/CAMI-challenge/OPAL) - they are approximately 20 MB.

bgruening commented on tools/cami_opal/cami_opal.xml:
```diff
@@ -0,0 +1,300 @@
+        ]]>
+    </command>
+    <inputs>
+        <param name="gold_standard_file" type="data" format="data" label="Gold standard file"
```
What format is that? The `format` should never be just `data`. Please change it to at least `txt` if its not any other specific format.

bgruening commented on tools/cami_opal/cami_opal.xml:
```diff
@@ -0,0 +1,300 @@
+    <inputs>
+        <param name="gold_standard_file" type="data" format="data" label="Gold standard file"
+            help="Input the gold standard file here. Format: CAMI Profiling Bioboxes." />
+        <param name="input_files" type="data" format="data" multiple="true" label="Input files"
```
same here
